### ISSUE TITLE
RFE: Easier Build Environment Configuration

### ISSUE THREAD

**Issue by ikaneshiro** [apptainer/singularity#2807]
There are a couple areas of image building related to environment handling that can be improved in order to ease the build process.

- Allowing `%post` to conditionally source the environment of the base image and potentially the `%environment` section as well.
- Allow `%test` to inherit the environment of `%post` to make testing at build time easier.

We have encountered some tricky behavior with how we source our environment that will need to be considered when investigating these options.

**Comment by HadrienG2** [comment]
I think that here, one key design question is, how much do we want Singularity to discriminate between the following environment variable usage scenarios?

1. Variables used during the %post section of a build and not needed afterwards.
2. Variables which should be made available to the %post section of dependent images.
3. Variables used during testing (which may vary depending on the kind of testing being carried out).
4. Variables used "at runtime", e.g. when a shell is opened into the output image.

Docker encourages using the same environment everywhere, which may result in somewhat unclean runtime environments. Singularity marks a clear distinction between 1/ and 4/, and currently has no good solution for 2/ and 3/. It seems to me that there is a trade-off between how many of these distinctions you want to make, and how complex the user interface for environment management in definition files will be.

As a small ergonomic tweak to the existing Singularity environment management concepts, you may also want to provide a shell function for the common `echo "export FOO=\"${FOO}\"" >> ${SINGULARITY_ENVIRONMENT}` pattern in %post sections.

**Comment by ikaneshiro** [comment]
@HadrienG2 reiterating some thoughts from my email on the mailing list so that they're tracked here as well. I think `2.` can be addressed by exposing an environment variable such as `SINGULARITY_BASE_ENVIRONMENT` that can be sourced in the post, e.g. `source $SINGULARITY_BASE_ENVIRONMENT`. I think point `3.` can be addressed in a similar fashion as `2.`

**Comment by Flamefire** [comment]
I just came across this as there is a major issue: Creating a definition file is hard.

TLDR: `singularity exec docker://someDockerImage myCommand` and `myCommand` executed in the following definition file should behave the same:
```
Bootstrap: docker
From: someDockerImage
%post
myCommand
```

This is currently not the case, as the environment from the source Docker container is included in the final image but not in the build environment. The [docu](https://sylabs.io/guides/3.3/user-guide/environment_and_metadata.html#environment) somewhat ambiguously states

> If you build a container from the Container Library or Docker Hub, the environment will be included with the container at build time.

Relevant use case:   
We usually advise our users to start from a base image in a writeable sandbox, shell inside it and execute all commands required for their installation. Then test their application and reiterate until it works. Afterwards (or during) copy all used commands inside a definition file. This saves a lot of time compared to rebuilding from scratch using a definition file which is iteratively enhanced.

For case 3 (tests) see also #3644

**Comment by GodloveD** [comment]
Hello Alexander!

Singularity used to source the files in the /.singularity.d/env during
build, but after extensive discussion it was decided to change this
behavior.  If you are interested in the gory details, the thread is here:

https://github.com/sylabs/singularity/issues/1053

As you can see by the age of the issue, this behavior was changed a long
time ago.

**Comment by Flamefire** [comment]
> Singularity used to source the files in the /.singularity.d/env during build, 

From there I see 

> I would agree the complete idea to run the %environment section already during build is wrong. 

with a quote from the manual which is still present (but modified):

> As of Singularity 2.3, you can add environment variables to be sourced when the container is used in the %environment section. 

However this issue here is different:

- The base environment from the boostrap image (given at `From:`) should be sourced at build time. There is no reason not to do that (that I can think of) and (I'm sure) the discussing in #1053 does not apply to this.
- The final environment should be source before executing tests, so that `singularity test` and the test after build behave the same (see #3644)

Reasoning is simple: If you start from a base image build with `Tool X` you want to use that in your build, especially now that multi-stage builds are supported (Even more suitable to start with an image including every possible build tool, build your stuff, then transfer to a slim image)

**Comment by Flamefire** [comment]
@GodloveD Bringing this up again: The missing environment *from the base container* is really confusing and annoying especially when developing "chains of containers".

When starting from a Docker image one can do `source /.singularity.d/env/10-docker*.sh`, which is IMO undocumented also a possibly very common use case. HPCCM (HPC Container Maker) does that by default.

But now I'm starting from a singularity image. In /.singularity.d/env I have `01-base.sh  10-docker2singularity.sh  90-environment.sh  94-appsbase.sh  95-apps.sh  99-base.sh  99-runtimevars.sh` and `SINGULARITY_ENVIRONMENT=/.singularity.d/env/91-environment.sh` What files am I supposed to source? How could a user ever supposed to know? And finally: It feels very wrong to go into a hidden "singularity-internal" directory.

Another argument: We generally recommend users to not write a definition file right away but shell into a writable (base) container, execute the commands there while recording them and once everything works put those into a definition file. Otherwise the process takes to much time as it starts from scratch everytime you made a mistake in the 100th step of installing something in the definition file.   
This workflow is broken if the base container environment differs significantly between the build and a shell.

**Comment by carterpeel** [comment]
Hello,

This is a templated response that is being sent out to all open issues.  We are working hard on 'rebuilding' the Singularity community, and a major task on the agenda is finding out what issues are still outstanding.


**Please consider the following:**

1. Is this issue a duplicate, or has it been fixed/implemented since being added?
2. Is the issue still relevant to the current state of Singularity's functionality?
3. Would you like to continue discussing this issue or feature request?


Thanks,
Carter

**Comment by pedroalvesbatista** [comment]
@ikaneshiro Still getting this problem ? Have this been solved already ? If yes, what work around have you followed and applied ?

We're looking into the issue carefully, soon will bring to community and discuss ways to better solve as well address this. Thankyou for keeping the interest in the subject.
